cpp/tests/utilities/column_utilities.cu

```diff
@@ -119,95 +130,201 @@ class corresponding_rows_not_equivalent {
-      std::vector<std::string> h_right_strings = to_strings(diff_table->get_column(1));
+// non-nested column types
+template <typename T, bool check_exact_equality>
+struct column_comparator_impl {
+  void operator()(column_view const& lhs,
+                  column_view const& rhs,
+                  bool print_all_differences,
+                  int depth)
+  {
+    using ComparatorType = std::conditional_t<check_exact_equality,
+                                              corresponding_rows_unequal,
+                                              corresponding_rows_not_equivalent>;
 
-      for (size_t i = 0; i < differences.size(); ++i) {
-        buffer << "lhs[" << differences[i] << "] = " << h_left_strings[i] << ", rhs["
-               << differences[i] << "] = " << h_right_strings[i] << std::endl;
-      }
+    auto d_lhs = cudf::table_device_view::create(table_view{{lhs}});
+    auto d_rhs = cudf::table_device_view::create(table_view{{rhs}});
 
-      EXPECT_EQ(differences.size(), size_t{0}) << buffer.str();
-    } else {
-      //
-      //  If there are differences, just display the first one
-      //
-      int index = differences[0];
+    // worst case - everything is different
+    thrust::device_vector<int> differences(lhs.size());
+
+    auto diff_iter = thrust::copy_if(thrust::device,
+                                     thrust::make_counting_iterator(0),
+                                     thrust::make_counting_iterator(lhs.size()),
+                                     differences.begin(),
+                                     ComparatorType(*d_lhs, *d_rhs));
+
+    // shrink back down
+    differences.resize(thrust::distance(differences.begin(), diff_iter));
+    print_differences(differences, lhs, rhs, print_all_differences, depth);
+  }
+};
 
-      auto diff_lhs = cudf::detail::slice(lhs, index, index + 1);
-      auto diff_rhs = cudf::detail::slice(rhs, index, index + 1);
+// forward declaration for nested-type recursion.
+template <bool check_exact_equality>
+struct column_comparator;
 
-      std::vector<std::string> h_left_strings  = to_strings(diff_lhs);
-      std::vector<std::string> h_right_strings = to_strings(diff_rhs);
+// specialization for list columns
+template <bool check_exact_equality>
+struct column_comparator_impl<list_view, check_exact_equality> {
+  void operator()(column_view const& lhs,
+                  column_view const& rhs,
+                  bool print_all_differences,
+                  int depth)
+  {
+    lists_column_view lhs_l(lhs);
+    lists_column_view rhs_l(rhs);
 
-      EXPECT_EQ(differences.size(), size_t{0})
-        << "first difference: "
-        << "lhs[" << index << "] = " << to_string(diff_lhs, "") << ", rhs[" << index
-        << "] = " << to_string(diff_rhs, "");
-    }
+    // using the row_equality_operator directly on a list column is a bad idea for several
+    // reasons:
+    // - at the moment, the row_equality_operator doesn't support lists
```

Comment:
Should the `row_equality_operator` support lists?
Sounds like it would _algorithmically terrible_ :-)
Does this mean overriding every operation that requires a sort of column?

Reply:
Yeah, this is an open question.  Sorts, groupbys, etc are definitely going to be tricky.  I talked with @jrhemstad  about this a bit.  The interesting thing about this is that :  if you use the Arrow format (as we are), this is a problem even for plain CPU implementations.  To figure out what the value of  "row" is requires recursive traversal of hierarchy underneath.   

Only way round it I can see is to flatten everything first, but that would be potentially explosive memory-wise.   Classic problem :)